### ISSUE TITLE
`Oj::Doc#fetch` gives an array instead of nil

### ISSUE THREAD

**Issue by BuonOmo** [ohler55/oj#662]
Hi @ohler55,

We've been using `Oj::Doc` for faster json handling recently, and we came on two issues, a bug (A) and what I think is a missing tool (B). Let me explain with an minimal example extracted from our codebase:

```ruby
def parse_route
  doc = Oj::Doc.open(body)
  return nil unless doc.fetch("code") == "Ok"


  doc.move("route")
  # (B) We are forced to load the whole json (if existing) just to know that
  # the key is present
  route = doc.fetch("0")
  # (A)`|| route.empty?` is here because of an Oj bug, returning an empty array
  return nil if route.nil? || route.empty?

  polyline = route["geometry"]
  distance = route["distance"].to_i
  duration = route["duration"].to_i
  RouteResult.new(
    distance: distance,
    duration: duration,
    polyline: polyline
  )
ensure
  doc&.close
end
```

For (A), the reproduction is quite obvious:

| code | result |
| -- | -- |
| `Oj::Doc.open('[]').fetch("/a")` | `[]` | 
| `Oj::Doc.open('[]').fetch("/0")` | `[]` | 
| `Oj::Doc.open('{}').fetch("/a")` | `{}` | 
| `Oj::Doc.open('{}').fetch("/0")` | `{}` | 
| `Oj::Doc.open('{}').fetch("/a/b")` | `{}` | 
| `Oj::Doc.open('[]').fetch("/0/a")` | `{}` | 

I think for all of these example we'd rather expect either `nil` or an error. I prefer nil, but it may conflict with a json actually having a `null` leaf (`{"a":null}`). Hence I'd rather let you decide on that

About (B) we could do something like below in our codebase.

```ruby
def path_exists?(doc, path)
  origin = doc.where?
  doc.move(path)
rescue ArgumentError
  false
else
  doc.move(origin)
  true
end
```

But it feels like it is the library responsibility, and doing it in c should be much faster. Hence I think a `Oj::Doc#exists?(path)` should be considered.

On last thing that has been bothering me a bit is `Oj::Doc#where?` which end with a `?`, hinting a boolean response, but always giving a string. I'd lean toward having a `Oj::Doc#path` alias, and documenting the alias, going towards a deprecation for some later time. This last bit is opinionated though (yet coming from me, you may not be that much surprised anymore ^^)

Of course I'd be glad to implement all of that on my free time, just letting the issue here for early feedback!

Cheers,
Ulysse

**Comment by ohler55** [comment]
A) agree, nil is probably best. Raising an exception seems a little too aggressive a response for a lookup.

B) Seems like a nice addition. I'll try to get something ready for review on both over the weekend if not sooner.

**Comment by ohler55** [comment]
The fetch-nil branch fixes the bug where fetching from an empty Hash or Array did not return nil. Note that if the Hash or Array were not empty nil was returned for a path that did not match any of the elements.

`Oj::Doc#where?` has been deprecated.

`Oj::Doc#exists?` has been added.

**Comment by BuonOmo** [comment]
FWIW, the branch looks good to me 😅 

Thank you for such dedication to your gem!

**Comment by ohler55** [comment]
I'll cut a release today then.

**Comment by ohler55** [comment]
Released v3.11.6
